### PR TITLE
remove addProgressCallback, add createDownloaderOperationWithUrl

### PR DESCRIPTION
**Pull request by gukemanbu**
### New Pull Request Checklist

* [x] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [x] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [x] I have searched for a similar pull request in the [project](https://github.com/rs/SDWebImage/pulls) and found none

* [x] I have updated this branch with the latest master to avoid conflicts (via merge from master or rebase)
* [x] I have added the required tests to prove the fix/feature I am adding
* [x] I have updated the documentation (if necessary)
* [x] I have run the tests and they pass
* [x] I have run the lint and it passes (`pod lib lint`)

This merge request fixes / reffers to the following issues: ...

### Pull Request Description

...

### CODE REVIEW

**Comment by gukemanbu** [comment]
@dreampiggy I don't think the original code is good design. The first thing I saw was createCallback code, while the main logic is in addProgressCallback method, which not conducive to reading.

**Comment by bpoplauschi** [comment]
I see your point @gukemanbu, but this change is breaking compatibility. We can talk about it, for a major release.

**Comment by bpoplauschi** [comment]
@gukemanbu I didn't properly look at your PR the last time, sorry for that.
Thanks for your interest in our project and the contributions you made.

Regarding this PR here, I agree with you. The code design was not that great, I think your's is much cleaner. I will look at this one more time and try to merge it.

**Comment by bpoplauschi** [comment]
I did merge it - so thanks @gukemanbu 
- I did not move the `cancel` method so it's easier to see the actual diff
- for the Downloader tests, I actually kept the original idea: check if passing a `nil` url will call the completion with all nils. Also added a check for the `createDownloaderOperationWithUrl:` non-nil result

**Comment by dreampiggy** [comment]
This feature, will not be available for 5.x branch. Now `master` and `5.x` are separate and not mergeable. If you wish to keep this design, please provide a new PR for 5.x branch as well.

**Comment by bpoplauschi** [comment]
@dreampiggy I think this change is not a big one. We will need to merge `master` into `5.x` anyway before the final release.
But if @gukemanbu is so kind to make a PR for the `5.x` branch, that would be appreciated.

**Comment by dreampiggy** [comment]
@bpoplauschi The diff is hard..I try to merge one, but however, this are many conflicts, nearlly 10+ files. 😅 

I suggest to view the diff result, but using git diff and copy the code instead...Especially any changes into `SDImageCoderXXX`, because they're be renamed and git will recognize it as a new file.

**Comment by bpoplauschi** [comment]
Hmm, I see only 6 files with conflict. I see d5b60c6 commit was a merge from `master` to `5.x` which you made (many thanks for that) on Jul 18th. So there are only a handful of commits since then.

**Comment by bpoplauschi** [comment]
I will look into this merge.

**Comment by dreampiggy** [comment]
@bpoplauschi You can try simulate #2399 #2409 together...Because these PR change the `SDImageCoder` and `SDWebImageDownloader`...

Anyway, it's just need time, maybe I can try provide a merge.

File: Tests/Tests/SDWebImageDownloaderTests.m [hunk 1]
```diff
@@ -20,10 +20,8 @@ @interface SDWebImageDownloader ()
 @property (assign, nonatomic, nullable) Class operationClass;
 @property (strong, nonatomic, nonnull) NSOperationQueue *downloadQueue;
 
-- (nullable SDWebImageDownloadToken *)addProgressCallback:(SDWebImageDownloaderProgressBlock)progressBlock
-                                           completedBlock:(SDWebImageDownloaderCompletedBlock)completedBlock
-                                                   forURL:(nullable NSURL *)url
-                                           createCallback:(SDWebImageDownloaderOperation *(^)(void))createCallback;
+- (SDWebImageDownloaderOperation *)createDownloaderOperationWithUrl:(nullable NSURL *)url options:(SDWebImageDownloaderOptions)options;
```
[the remark this answers is not in the document]
Comment (bpoplauschi) [reply Tests/Tests/SDWebImageDownloaderTests.m]: It's fixed in the merge commit. Here we still see the diff with the original branch, without the conflicts fixed.

File: Tests/Tests/SDWebImageDownloaderTests.m [hunk 2]
```diff
@@ -119,16 +117,8 @@ - (void)test06ThatUsingACustomDownloaderOperationWorks {
...
-    [[SDWebImageDownloader sharedDownloader] addProgressCallback:nil completedBlock:^(UIImage * _Nullable image, NSData * _Nullable data, NSError * _Nullable error, BOOL finished) {
-        if (!image && !data && !error) {
-            [expectation fulfill];
-        } else {
-            XCTFail(@"All params should be nil");
-        }
-    } forURL:nil createCallback:nil];
-    [self waitForExpectationsWithTimeout:0.5 handler:nil];
+- (void)test07ThatCreateDownloaderOperationWithNilUrl {
+    [[SDWebImageDownloader sharedDownloader] createDownloaderOperationWithUrl:nil options:0];
```
[the remark this answers is not in the document]
Comment (bpoplauschi) [reply Tests/Tests/SDWebImageDownloaderTests.m]: I've also fixed this in the merge commit. Check that. 676a4b5

**Comment by dreampiggy** [comment]
@bpoplauschi I merge the lastest master to 5.x.  Actually the changes is only about this PR (Another changes which apply to FLAnimatedImage, is already fixed in [SDWebImageFLPlugin](https://github.com/SDWebImage/SDWebImageFLPlugin)).

Actually, just some code movement, this PR :)

**Comment by bpoplauschi** [comment]
Yeap. Thanks @dreampiggy for doing the merge. I was doing it myself, but that's ok.

**Comment by gukemanbu** [comment]
Thank you all!

**Comment by bpoplauschi** [comment]
@gukemanbu did you see my invitation addressed to you above?

**Comment by gukemanbu** [comment]
> @gukemanbu did you see my invitation addressed to you above?

@bpoplauschi I'm too busy these days to see it in time, Sorry for the late reply, I'm very excited to become a maintainer.

**Comment by bpoplauschi** [comment]
Awesome @gukemanbu. I sent you an invitation to the organization.

**Comment by gukemanbu** [comment]
@bpoplauschi  Thank you very much，I have accepted the invitation。

**Pull request merged**